generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,382 @@
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = "*" | "**" | LITERAL | Variable ;
+//	Variable = "{" FieldPath [ "=" Segments ] "}" ;
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+//
+// The specific notation is not defined, but it seems inspired by
+// [Backus-Naur Form]. In this notation, `{ ... }` allows repetition.
+//
+// The documentation goes on to say:
+//
+//	A variable template must not contain other variables.
+//
+// So the grammar is better defined by:
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = Variable | PlainSegment;
+//	PlainSegment  = "*" | "**" | LITERAL ;
+//	Variable = "{" FieldPath [ "=" PlainSegments ] "}" ;
+//	PlainSegments = PlainSegment { "/" PlainSegment };
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+//
+// Neither "IDENT" nor "LITERAL" are defined. From context we can infer that
+// IDENT must be a valid proto3 identifier, so matching the regular expression
+// `[A-Za-z][A-Za-z0-9_]*`. Likewise, we can infer that LITERAL must be a path
+// segment in a URL. [RFC 3986] provides a definition for these, which we
+// summarize as:
+//
+// LITERAL     = pchar { pchar }
+// pchar       = unreserved | pct-encoded | sub-delims | ":" | "@"
+// unreserved  = ALPHA | DIGIT | "-" | "." | "_" | "~"
+// pct-encoded = "%" HEXDIG HEXDIG
+// sub-delims  = "!" | "$" | "&" | "'" | "(" | ")"
+//
+//	| "*" | "+" | "," | ";" | "="
+//
+// ALPHA       = [A-Za-z]
+// DIGIT       = [0-9]
+// HEXDIG      = [0-9A-Fa-f]
+//
+// [RFC 3986]: https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
+// [Backus-Naur Form]: https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+
+func Parse(pathTemplate string) (*PathTemplate, error) {
+	return parsePathTemplate(pathTemplate)
+}
+
+// ParseSegments flattens the result of Parse into a slice of api.PathSegment,
+// ignoring variable values and match (* and **) segments.
+// TODO: This function is a temporary shim to allow the existing tests to pass.
+func ParseSegments(pathTemplate string) ([]api.PathSegment, error) {
+	path, err := parsePathTemplate(pathTemplate)
+	if err != nil {
+		return nil, err
+	}
+	var segments []api.PathSegment
+	for _, s := range path.Segments {
+		segment := api.PathSegment{}
+		if s.Literal != nil {
+			literal := string(*s.Literal)
+			segment.Literal = &literal
+		} else if s.Variable != nil {
+			ids := make([]string, len(s.Variable.FieldPath))
+			for i, id := range s.Variable.FieldPath {
+				ids[i] = string(*id)
+			}
+			fieldPath := strings.Join(ids, ".")
+			segment.FieldPath = &fieldPath
+		}
+		segments = append(segments, segment)
+	}
+
+	if path.Verb != nil {
+		verb := string(*path.Verb)
+		segments = append(segments, api.PathSegment{
+			Verb: &verb,
+		})
+	}
+	return segments, nil
+}
+
+// PathTemplate represents the structure in Go.
+type PathTemplate struct {
```

Comment:
looks like none of these types are used outside the package. They should use be unexported and use lowercase.  If you are planning to export them in a follow up PR leave them as they are.

Reply:
Added #557 to address this, my intention is to use these in the future.